lib/ddtrace/profiling/pprof/builder.rb

```diff
@@ -0,0 +1,181 @@
+                  group_values[i] = group_value + values[i]
+                end
+              else
+                # Add new group
+                event_groups[key] = [event, values]
+              end
+            end
+          end
+
+          event_groups.collect do |_group_key, group|
+            yield(
+              # Event
+              group[0],
+              # Values
+              group[1]
+            )
+          end
+        end
+
+        def event_group_key(event)
+          raise NotImplementedError
+        end
+
+        def build_sample_types
+          raise NotImplementedError
+        end
+
+        def build_value_type(type, unit)
+          Perftools::Profiles::ValueType.new(
+            type: string_table.fetch(type),
+            unit: string_table.fetch(unit)
+          )
+        end
+
+        def build_sample(event, values)
+          raise NotImplementedError
+        end
+
+        def build_sample_values(event)
+          raise NotImplementedError
+        end
+
+        def build_locations(backtrace_locations, length)
+          locations = backtrace_locations.collect do |backtrace_location|
+            @locations.fetch(
+              # Filename
+              backtrace_location.path,
+              # Line number
+              backtrace_location.lineno,
+              # Function name
+              backtrace_location.base_label,
+              # Build function
+              &method(:build_location)
+            )
+          end
+
+          omitted = length - backtrace_locations.length
+
+          # Add placeholder stack frame if frames were truncated
+          if omitted > 0
+            desc = omitted == 1 ? DESC_FRAME_OMITTED : DESC_FRAMES_OMITTED
+            locations << @locations.fetch(
+              '',
+              0,
+              "#{omitted} #{desc}",
+              &method(:build_location)
+            )
+          end
+
+          locations
+        end
+
+        def build_location(id, filename, line_number, function_name = nil)
+          Perftools::Profiles::Location.new(
+            id: id,
+            line: [build_line(
+              @functions.fetch(
+                filename,
+                function_name,
+                &method(:build_function)
+              ).id,
+              line_number
+            )]
+          )
+        end
+
+        def build_line(function_id, line_number)
+          Perftools::Profiles::Line.new(
+            function_id: function_id,
+            line: line_number
+          )
+        end
+
+        def build_function(id, filename, function_name)
+          Perftools::Profiles::Function.new(
+            id: id,
+            name: @string_table.fetch(function_name),
+            filename: @string_table.fetch(filename)
+          )
+        end
+
+        def build_mappings
+          [
+            Perftools::Profiles::Mapping.new(
+              id: 1,
+              filename: @string_table.fetch($PROGRAM_NAME)
```

Comment:
What are we using `$PROGRAM_NAME` for at the end?
I ask because for a Rails app I get `$PROGRAM_NAME => "bin/rails"` (for `bin/rails s`) or `$PROGRAM_NAME => "spring app    | rails5_2_3 | started 0 secs ago | development mode"` (for `bin/rails c`).

Reply:
We're using this to label the profile as being associated with a particular process.

But I had no idea that Rails would modify this such that it'd include something mutable like `started 0 secs ago`, so that's a great catch.

Reply:
So it is mutable, but I'm unable to replicate your results...

For both `bin/rails c` and `bin/rails s` I get:

```
0: bin/rails
PROGRAM_NAME: bin/rails
__FILE__: /app/config/initializers/datadog.rb # (whatever the current file is)
```